src/Network/Wai/Auth/Tools.hs

```diff
@@ -0,0 +1,46 @@
+{-# LANGUAGE BangPatterns #-}
+module Network.Wai.Auth.Tools
+  ( encodeKey
+  , decodeKey
+  , toLowerUnderscore
+  , getValidEmail
+  ) where
+
+import qualified Data.ByteString        as S
+import           Data.ByteString.Base64 as B64
+import           Data.Char              (isLower, toLower)
+import           Data.Foldable          (foldr')
+import           Data.Serialize         (Get, get, put, runGet, runPut)
+import qualified Data.Text              as T
+import           Web.ClientSession      (Key)
+
+
+-- | Decode a `Key` that is in a base64 encoded serialized form
+decodeKey :: Monad m => S.ByteString -> m Key
```

Comment:
The `fail` function is generally despised :). For many monads, it is implemented as `error`, which means we can't rely on it being total. By using `MonadThrow` and `throwM` instead, we're guaranteed totality.